src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+				                                                   (int)((double)(screen_x % VIDEO_WIDTH) * stepx));
+			}
+		}
+	}
+	return 0;
+}
+
+uint8_t dfs500_state::pit_r(offs_t offset)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	return m_pit->read((offset>>1) & 3); // addressed by CPU's address bits AB2 and AB1
+}
+
+void dfs500_state::pit_w(offs_t offset, uint8_t data)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	m_pit->write((offset>>1) & 3, data); // addressed by CPU's address bits AB2 and AB1
+}
+
+uint8_t dfs500_state::cpanel_pit_r(offs_t offset)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	return m_cpanel_pit->read((offset>>1) & 3); // addressed by CPU's address bits AB2 and AB1
+}
+
+void dfs500_state::cpanel_pit_w(offs_t offset, uint8_t data)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	m_cpanel_pit->write((offset>>1) & 3, data); // addressed by CPU's address bits AB2 and AB1
+}
+
+uint8_t dfs500_state::pic_r(offs_t offset)
+{
+	// PD71059C (Programmable Interrupt Controller)
+	return m_pic->read((offset>>1) & 1); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::pic_w(offs_t offset, uint8_t data)
+{
+	// PD71059C (Programmable Interrupt Controller)
+	m_pic->write((offset>>1) & 1, data); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::rombank1_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank1->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::rombank2_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank2->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::input_select_w(offs_t offset, uint8_t data)
+{
+	// Selects sources of video input on the AD-76 board.
+	m_input_sel_A = (data >> 3) & 0x7;
+	m_input_sel_B = data & 0x7;
+}
+
+void dfs500_state::WA0_w(offs_t offset, uint16_t data)
+{
+	m_effectcpu_latch16 = data;
+	m_TOC = true;
+}
+
+uint16_t dfs500_state::RA0_r(offs_t offset)
+{
+	return m_maincpu_latch16;
+	m_TOE = false;
+}
+
+uint16_t dfs500_state::RA1_r(offs_t offset)
+{
+	// "TEST, 1, OPT2, OPT1, RFLD, VD, TOC, TOE"
+	uint8_t value = 0x40;
+	// FIXME! Add other signals.
+	if (m_TOC) value |= 0x02;
+	if (m_TOE) value |= 0x01;
+	value |= ((ioport("DSW_S3")->read() & 0x0f) << 8); // (Unknown)
+	return value;
+}
+
+void dfs500_state::WB0_w(offs_t offset, uint16_t data)
+{
+	m_maincpu_latch16 = data;
+	m_TOE = true;
+}
+
+uint16_t dfs500_state::RB0_r(offs_t offset)
+{
+	return m_effectcpu_latch16;
+	m_TOC = false;
+}
+
+uint8_t dfs500_state::RB1_r(offs_t offset)
+{
+	//"TEST, OPT2, OPT1, VD, T2, T1, TOE, TOC"
+	uint8_t value = 0;
+	// FIXME! Add other signals.
+	if (m_TOE) value |= 0x02;
+	if (m_TOC) value |= 0x01;
+	return value;
+}
+
+uint8_t dfs500_state::RB2_r(offs_t offset)
+{
+	uint8_t value = 0;
+	value |= ((ioport("DSW_S1")->read() & 0x0f) << 4); // (Editing Control Unit Select)
+	value |= (ioport("DSW_S2")->read() & 0x0f);        // (Freeze Timing)
```

Comment:
Don’t do tag lookups on the hot path - they’re expensive, and you get no validation.

Reply:
done